### ISSUE TITLE
Best practices concerning referential transparency in `equinox`

### ISSUE THREAD

**Issue by BenjaminDAnjou** [patrick-kidger/equinox#1236]
In this [issue](https://github.com/patrick-kidger/equinox/issues/1231), the discussion drifted towards referential transparency in `equinox`. This left me with some questions that are better asked separately.

Consider the minimum working example with 4 different options for the `__init__` method:

```python
# CODE SAMPLE 1

import equinox as eqx
import jax.numpy as jnp
from jax.numpy.fft import fft

class TimeSeries(eqx.Module):

    time_series: ArrayLike
    spectrum: ArrayLike

    def __init__(self, time_series):

        self.time_series = time_series
        self.spectrum = fft(time_series) # OPTION 1
        self.spectrum = fft(self.time_series) # OPTION 2
        self.spectrum = self.get_spectrum() # OPTION 3
        # OPTION 4: None of the above; avoid storing spectrum entirely

    def get_spectrum(self):
        return fft(self.time_series)
    
series = jnp.ones((10,), dtype=float)
TimeSeries(series)
```

This class stores a time series and its spectrum. The spectrum is clearly a function of the time series. So conceptually, the fields of the class don't "represent something independent of all the others", which was discouraged in the previous issue.

**Which of the options listed in the example, if any, are good practice in `equinox`?**

And if it turns out to be `OPTION 4`, and we **must** avoid storing the `spectrum` field, how can we ever justify composing classes like this:

```python
# CODE SAMPLE 2

class ClassA(eqx.Module):

    A: ArrayLike

    def __init__(self, A):
        self.A = A

class ClassB(eqx.Module):

    B: ArrayLike
    _obj_A: ClassA
    
    def __init__(self, B):
        self.B = B
        self._obj_A = ClassA(B**2)
```

Here, the field `self._obj_A` is clearly not represent something independent from the field `self.B` since they are initialized from the same array. And yet this is the form that is recommended by the [abstract/final pattern](https://docs.kidger.site/equinox/pattern/).

**Comment by patrick-kidger** [comment]
So the question really boils down to "if you pass this object across a `jax.grad` boundary, do you want the fields to get gradients independently of each other?"

If you want separate gradients, then they should be separate fields. If you want the gradient of one 'derived field' to actually just accumulate into the 'original field', then they must not be separate fields.

Since accumulation is the most common scenario, I therefore generally recommend each field being independent, which is your 'option 4'. But if you want different behaviour then you should feel free to use any of the (equivalent) options 1-3 instead.

---

For your final question around composition, the answer is the same. If you want separate gradients then what you have is fine. If you want one to accumulate into the other, then the same rules apply: the 'derived' thing (here, the entire `ClassA(B**2)`) should be moved from initialisation and should be created in the forward pass instead.
